rclpy/rclpy/executors.py

```diff
@@ -0,0 +1,290 @@
+            for h in handles:
+                _rclpy.rclpy_wait_set_add_entity(
+                    entity, wait_set, h.__getattribute__(handle_name)
+                )
+        _rclpy.rclpy_wait_set_clear_entities('guard_condition', wait_set)
+        _rclpy.rclpy_wait_set_add_entity('guard_condition', wait_set, sigint_gc)
+        _rclpy.rclpy_wait_set_add_entity('guard_condition', wait_set, self._guard_condition)
+
+        # Wait for something to become ready
+        _rclpy.rclpy_wait(wait_set, timeout_nsec)
+
+        # get ready entities
+        sub_ready_list = _rclpy.rclpy_get_ready_entities('subscription', wait_set)
+        guard_condition_ready_list = _rclpy.rclpy_get_ready_entities('guard_condition', wait_set)
+        timer_ready_list = _rclpy.rclpy_get_ready_entities('timer', wait_set)
+        client_ready_list = _rclpy.rclpy_get_ready_entities('client', wait_set)
+        service_ready_list = _rclpy.rclpy_get_ready_entities('service', wait_set)
+
+        # Check sigint guard condition
+        if sigint_gc_handle in guard_condition_ready_list:
+            raise KeyboardInterrupt
+
+        # Process ready entities one node at a time
+        for node in nodes:
+            for tmr in [t for t in node.timers if t.timer_pointer in timer_ready_list]:
+                # Check that a timer is ready to workaround rcl issue with cancelled timers
+                if _rclpy.rclpy_is_timer_ready(tmr.timer_handle):
+                    if tmr.callback_group.can_take(tmr):
+                        handler = self._make_handler(tmr, self._take_timer, self._execute_timer)
+                        yield handler, tmr, node
+
+            for sub in [s for s in node.subscriptions if s.subscription_pointer in sub_ready_list]:
+                if sub.callback_group.can_take(sub):
+                    handler = self._make_handler(
+                        sub, self._take_subscription, self._execute_subscription)
+                    yield handler, sub, node
+
+            for client in [c for c in node.clients if c.client_pointer in client_ready_list]:
+                if client.callback_group.can_take(client):
+                    handler = self._make_handler(client, self._take_client, self._execute_client)
+                    yield handler, client, node
+
+            for srv in [s for s in node.services if s.service_pointer in service_ready_list]:
+                if srv.callback_group.can_take(srv):
+                    handler = self._make_handler(srv, self._take_service, self._execute_service)
+                    yield handler, srv, node
+
+
+class SingleThreadedExecutor(Executor):
+    """Runs callbacks one at a time in the thread which calls spin()."""
+
+    def __init__(self):
+        super().__init__()
+
+    def spin_once(self, timeout_sec=None):
+        for handler, entity, node in self.wait_for_ready_callbacks(
+                timeout_sec=timeout_sec):
+            handler()
+            break
```

Comment:
Since the `for` loop always break this could just get the `next` from the generator.

Same below.

Reply:
Fixed in 073146983d5be98509feaef4841887dc0d3359a5